devito/operator.py

```diff
@@ -150,62 +157,76 @@ def arguments(self, *args, **kwargs):
+        f_args = [(name, f) for name, f in arguments.items()
+                  if isinstance(f, SymbolicData)]
+        for fname, f in f_args:
+            arguments[fname] = self._arg_data(f)
             shape = self._arg_shape(f)
 
             # Ensure data dimensions match symbol dimensions
             for i, dim in enumerate(f.indices):
-                # Infer open loop limits
-                if dim.size is None:
-                    # First, try to find dim size in kwargs
-                    if dim.name in kwargs:
-                        dim_sizes[dim] = kwargs[dim.name]
-
-                    if dim in dim_sizes:
-                        # Ensure size matches previously defined size
-                        if not dim.is_Buffered:
-                            assert dim_sizes[dim] <= shape[i]
+                # We don't need to check sizes for buffered dimensions
+                # against data shapes, all we need is the size of the parent.
+                if dim.is_Buffered:
+                    continue
+
+                # Check data sizes for dimensions with a fixed size
+                if dim.size is not None:
+                    if not shape[i] <= dim.size:
+                        error('Size of data argument for %s is greater than the size '
+                              'of dimension %s: %d' % (fname, dim.name, dim.size))
+                        raise InvalidArgument('Wrong data shape encountered')
                     else:
-                        # Derive size from grid data shape and store
-                        dim_sizes[dim] = shape[i]
-                else:
-                    if not isinstance(dim, BufferedDimension):
-                        assert dim.size == shape[i]
+                        continue
 
-        # Ensure parent for buffered dims is defined
-        buf_dims = [d for d in dim_sizes if d.is_Buffered]
-        for dim in buf_dims:
-            if dim.parent not in dim_sizes:
-                dim_sizes[dim.parent] = dim_sizes[dim]
+                if dim_sizes[dim.name] is None:
+                    # We haven't determined the size of this dimension yet,
+                    # try to infer it from the data shape.
+                    dim_sizes[dim.name] = shape[i]
+                else:
+                    # We know the dimension size, check if data shape agrees
+                    if not dim_sizes[dim.name] <= shape[i]:
+                        error('Size of dimension %s was determined to be %d, '
+                              'but data for symbol %s has shape %d.'
+                              % (dim.name, dim_sizes[dim.name], fname, shape[i]))
+                        raise InvalidArgument('Wrong data shape encountered')
+
+        # Make sure we have defined all buffered dimensions and their parents,
+        # even if they are not explicitly given or used.
+        d_args = [d for d in arguments.values() if isinstance(d, Dimension)]
```

Comment:
Ok, so I'm not sure I understand you comment completely. `dim_sizes` is now `name->value`, whereas is was `object->value` before. The reason for that change is, so that we can easily override via `kwargs`, without having to derive the object from the name. The reason `name->object` won't work is that the `Dimension` object doesn't (and I argue shouldn't) store the particular value for a single invocation, since on the next invocation that value might be different. Imagine `op(time=5); do_some_checkpointing; op(time=3)`. As a corollary, when we actually need the `Dimension` object, we pull it from `arguments` based on name - but I guess we could do the same for `kwargs` replacement and make `dim_sizes` `object->value` again. I'm not sure which way is best, tbh...